packages/backend/src/server/activitypub.ts

```diff
@@ -224,4 +225,24 @@ router.get('/likes/:like', async ctx => {
 	setResponseType(ctx);
 });
 
+// follow
+router.get('/follows/:follower/:followee', async ctx => {
+	// This may be used before the follow is completed, so we do not
+	// check if the following exists.
+
+	const [follower, followee] = await Promise.all([
+		Users.findOneBy({ id: ctx.params.follower }),
+		Users.findOneBy({ id: ctx.params.followee }),
+	]);
+
+	if (follower == null || followee == null) {
+		ctx.status = 404;
+		return;
+	}
+
+	ctx.body = renderActivity(renderFollow(follower, followee));
+	ctx.set('Cache-Control', 'public, max-age=180');
+	setResponseType(ctx);
+});
```

Comment:
うーん、インスタンスに保存するユーザーのパターンなら、どんなパターンでも作らされてしまうのが少し気になってる。

Reply:
Yes, it will be technically possible to create any pair of users stored in the database.

As I have said above already, being able to create this means little: Misskey uses follow requests to track whether a follow was requested by a local user. Before accepting a follow, Misskey checks that a matching follow request exists. If it does not exist, nothing happens.

Similarly, using this ID to create a fraudulent Activity should also not work because the actor that made the signature will not be the same as the `actor` of the `Follow` activity. This would only be a concern if the private key becomes public, which is in my understanding outside the threat model of Misskey.

To ensure this is not abused I think it does make sense to check that the follower is a local user and the followee is not a local user by adding those parameters to the queries.

If this is not what you meant, please be more specific.

Reply:
まあ、そうは思うわ…

やっぱり、HTTP Resolver経由で取得するのが引っかかる。

Activity endpointは公開せずに /packages/backend/src/remote/activitypub/kernel/accept/follow.ts などでURLパースでDB解決するとかしたい。
Related #8688

Reply:
The resolve of the accepted object happens in `packages/backend/src/remote/activitypub/kernel/accept/index.ts`, and there a `Resolver` is used. Parsing the URL instead of making a HTTP request is the job of `DbResolver`. I am not sure if I can replace `Resolver` with a `DbResolver` there without breaking behaviour.

If it is possible, a new `getFollowFromApId` function could be added to `DbResolver` to be used instead, since misskey only seems to be able to accept `Follow` objects there anyway.

Reply:
I was just thinking about it.

In the case of Undo, there can be Follow and Block, but since it is from local, it can be identified by URL Path.

Reply:
As I said before, I think this should be handled in the issue you linked, because in my opinion, there needs to be a way to automatically fetching the right type of object locally, as I wrote in <https://github.com/misskey-dev/misskey/issues/8688#issuecomment-1128811932>.
An additional complexity is that the kernel always uses ActivityPub objects, so use of `DbResolver` might be inappropriate because it returns database objects.